### ISSUE TITLE
certificate verify failed

### ISSUE THREAD

**Issue by misrn** [kubernetes-client/python#1017]
python 3.7
client 10.0.1

    from kubernetes import client, config
    config.kube_config.load_kube_config(config_file="F:\\config.yaml")
    class Kubernetes:

        def __init__(self):
            self.Connect = client.CoreV1Api()

        def list_namespaces(self):
            data = []
            for ns in self.Connect.list_namespace().items:
                data.append(ns)
            return data
    k = Kubernetes()
    print(k.list_namespaces())

urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='172.18.56.103', port=6443): Max retries exceeded with url: /api/v1/namespaces (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get issuer certificate (_ssl.c:1056)')))

Can't verify the certificate. What should I do?

**Comment by roycaihw** [comment]
which OS are you running? SSL verification failures usually heavily depends on your environment setup. Please check out the SSL troubleshooting in readme: https://github.com/kubernetes-client/python#troubleshooting

you could also try disabling SSL verification if you're doing some testing. Ref https://github.com/kubernetes-client/python-base/issues/177

**Comment by micw523** [comment]
Looking at his file path I think it’s some version of Windows

**Comment by nettyxiong** [comment]
I miss the same problem. It seems that the packages of requirements.txt is statisfied. And I find that use python3 is working but not with python2.
The enviroment is as follows:
CentOS 7.3
```bash
[root@VM_8_53_centos /data/code/python]# python -c "import ssl; print ssl.OPENSSL_VERSION"
OpenSSL 1.0.1e-fips 11 Feb 2013
```

```bash
[root@VM_8_53_centos` /data/code/python]# pip freeze -r requirements.txt
certifi==2019.11.28
six==1.9.0
python-dateutil==2.8.1
setuptools==42.0.2
Requirement file contains pyyaml>=3.12, but that package is not installed
google-auth==1.10.0
ipaddress==1.0.23
websocket-client==0.57.0
requests==2.6.0
requests-oauthlib==1.3.0
urllib3==1.25.7
## The following requirements were added by pip freeze:
Babel==0.9.6
backports.ssl-match-hostname==3.4.0.2
cachetools==4.0.0
chardet==2.2.1
Cheetah==2.4.4
cloud-init==17.1
configobj==4.7.2
decorator==3.4.0
iniparse==0.4
iotop==0.6
IPy==0.75
javapackages==1.0.0
Jinja2==2.7.2
jsonpatch==1.2
jsonpointer==1.9
jsonschema==2.3.0
kitchen==1.1.1
kubernetes===11.0.0-snapshot
lxml==3.2.1
M2Crypto==0.21.1
Markdown==2.4.1
MarkupSafe==0.11
nose==1.3.0
numpy==1.7.1
oauthlib==3.1.0
perf==0.1
pexpect==2.3
Pillow==2.0.0
policycoreutils-default-encoding==0.1
prettytable==0.7.2
pyasn1==0.4.8
pyasn1-modules==0.2.7
pycrypto==2.6.1
pycurl==7.19.0
Pygments==1.4
pygobject==3.14.0
pygpgme==0.3
pyliblzma==0.5.3
pyparsing==1.5.6
python-ldap==2.4.15
pyudev==0.15
pyxattr==0.5.1
PyYAML==5.2
rsa==4.0
seobject==0.1
sepolicy==1.1
slip==0.4.0
slip.dbus==0.4.0
urlgrabber==3.10
yum-metadata-parser==1.1.4
```

**Comment by ikogan** [comment]
I'm having this issue on Fedora 31. The CA is in my system trust anchors, Chrome likes it, kubectl likes it. I've tried it with the `certificate-authority-data` in my kubeconfig and not in my kubeconfig. In both cases, kubectl likes it but the Python module does not:

OpenSSL version is `OpenSSL 1.1.1d FIPS  10 Sep 2019`
Python version is 3.7.5.
```
# pip install kubernetes
Requirement already satisfied: kubernetes in /usr/local/lib/python3.7/site-packages (10.0.1)
Requirement already satisfied: six>=1.9.0 in /usr/lib/python3.7/site-packages (from kubernetes) (1.12.0)
Requirement already satisfied: websocket-client!=0.40.0,!=0.41.*,!=0.42.*,>=0.32.0 in /usr/local/lib/python3.7/site-packages (from kubernetes) (0.57.0)
Requirement already satisfied: requests in /usr/lib/python3.7/site-packages (from kubernetes) (2.22.0)
Requirement already satisfied: urllib3>=1.24.2 in /usr/lib/python3.7/site-packages (from kubernetes) (1.25.7)
Requirement already satisfied: google-auth>=1.0.1 in /usr/local/lib/python3.7/site-packages (from kubernetes) (1.10.0)
Requirement already satisfied: pyyaml>=3.12 in /usr/lib64/python3.7/site-packages (from kubernetes) (5.1.2)
Requirement already satisfied: requests-oauthlib in /usr/local/lib/python3.7/site-packages (from kubernetes) (1.3.0)
Requirement already satisfied: certifi>=14.05.14 in /usr/local/lib/python3.7/site-packages (from kubernetes) (2019.9.11)
Requirement already satisfied: setuptools>=21.0.0 in /usr/lib/python3.7/site-packages (from kubernetes) (41.2.0)
Requirement already satisfied: python-dateutil>=2.5.3 in /usr/lib/python3.7/site-packages (from kubernetes) (2.8.0)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/lib/python3.7/site-packages (from requests->kubernetes) (3.0.4)
Requirement already satisfied: idna<2.9,>=2.5 in /usr/lib/python3.7/site-packages (from requests->kubernetes) (2.8)
Requirement already satisfied: pyasn1-modules>=0.2.1 in /usr/local/lib/python3.7/site-packages (from google-auth>=1.0.1->kubernetes) (0.2.8)
Requirement already satisfied: rsa<4.1,>=3.1.4 in /usr/local/lib/python3.7/site-packages (from google-auth>=1.0.1->kubernetes) (4.0)
Requirement already satisfied: cachetools<5.0,>=2.0.0 in /usr/local/lib/python3.7/site-packages (from google-auth>=1.0.1->kubernetes) (4.0.0)
Requirement already satisfied: oauthlib>=3.0.0 in /usr/local/lib/python3.7/site-packages (from requests-oauthlib->kubernetes) (3.1.0)
Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /usr/local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth>=1.0.1->kubernetes) (0.4.8)
```

**Comment by degerrit** [comment]
Had the same issue here, but found a solution for me - which may help some.

Previously, even though `curl` worked fine for me - python 3.5 & client `kubernetes==11.0.0` threw up an `SSL: CERTIFICATE_VERIFY_FAILED` error when using the exact same bearer token and CA file (which in my case is simply the public, Rancher self-signed, Kube API server cert itself).

SSL connection worked for `curl`:
`# curl --noproxy chq20  --url https://chq20:6443/healthz --cacert ./check_qa.pem --header 'Authorization: Bearer deadbeef' --max-time 1.0`

But SSL connection failed as above, using the example code in `python/examples/remote_cluster.py` and the same token/CA file as worked above.

(I had also checked that python is connecting to the correct host, and not our corporate proxy -which is another common problem here- using `strace -e trace=network python3` : connection destination looks good)

In the end, I found `openssl s_client -connect chq20:6443 -CAfile ./check_qa.pem` indicated a few errors (unlike @ikogan, thanks for the tip though):
```
CONNECTED(00000003)
depth=0 CN = kube-apiserver
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = kube-apiserver
verify error:num=21:unable to verify the first certificate
verify return:1
```

So in my case the problem was using an unexpected CA file (and assuming all ssl client implementations behave the same!), and therefore an inability for more strict SSL clients to check the complete trust chain (which the API server cert does refer to). 

I fetched the right Rancher CA from a Rancher container in the right cluster `/etc/kubernetes/ssl/kube-ca.pem` and am now using that instead.
